parsing/dialects/dialect1/parserImpls.ftl

```diff
@@ -2382,3 +2382,163 @@ SqlCreateJoinIndex SqlCreateJoinIndex() :
+        {
+            nodes.add(e);
+        }
+    )*
+    [
+        <COMMA>
+        (
+            LOOKAHEAD(3)
+            <NO> <CASE> <OR> <UNKNOWN>
+            {
+                extraPartitionOption =
+                    SqlPartitionByNoneUnknown.NONE_OR_UNKNOWN;
+            }
+        |
+            LOOKAHEAD(3)
+            <NO> <CASE> <COMMA> <UNKNOWN>
+            {
+                extraPartitionOption =
+                    SqlPartitionByNoneUnknown.NONE_COMMA_UNKNOWN;
+            }
+        |
+            <NO> <CASE>
+            { extraPartitionOption = SqlPartitionByNoneUnknown.NONE; }
+        |
+            <UNKNOWN>
+            { extraPartitionOption = SqlPartitionByNoneUnknown.UNKNOWN; }
+        )
+    ]
+    <RPAREN>
+    {
+        return new SqlCaseN(getPos(), nodes, extraPartitionOption);
+    }
+}
+
+SqlRangeN RangeN() :
+{
+    final SqlNode testExpression;
+    SqlNode range;
+    SqlNode startLiteral = null;
+    boolean startAsterisk = false;
+    SqlNode endLiteral = null;
+    boolean endAsterisk = false;
+    SqlNode eachSizeLiteral = null;
+    final SqlNodeList rangeList = new SqlNodeList(getPos());
+    SqlPartitionByNoneUnknown extraPartitionOption = null;
+}
+{
+    <RANGE_N>
+    <LPAREN>
+    testExpression = CompoundIdentifier()
+    <BETWEEN>
+    (
+        <STAR> { startAsterisk = true; }
+        [
+            <AND>
+            (
+                <STAR> { endAsterisk = true; }
+            |
+                endLiteral = Literal()
+            )
+        ]
+    |
+        startLiteral = Literal()
+        [
+            <AND>
+            (
+                <STAR> { endAsterisk = true; }
+            |
+                endLiteral = Literal()
+            )
+        ]
+        [
+            <EACH>
+            eachSizeLiteral = Literal()
+        ]
+    )
+
+    {
+        rangeList.add(new SqlRangeNStartEnd(getPos(), startLiteral, endLiteral,
+            eachSizeLiteral, startAsterisk, endAsterisk));
+    }
+    (
+        <COMMA>
+        range = RangeNStartEnd()
+        { rangeList.add(range); }
+    )*
+    [
+        <COMMA>
+        (
+            LOOKAHEAD(3)
+            <NO> <RANGE> <OR> <UNKNOWN>
+            {
+                extraPartitionOption =
+                    SqlPartitionByNoneUnknown.NONE_OR_UNKNOWN;
+            }
+        |
+            LOOKAHEAD(3)
+            <NO> <RANGE> <COMMA> <UNKNOWN>
+            {
+                extraPartitionOption =
+                    SqlPartitionByNoneUnknown.NONE_COMMA_UNKNOWN;
+            }
+        |
+            <NO> <RANGE>
+            { extraPartitionOption = SqlPartitionByNoneUnknown.NONE; }
+        |
+            <UNKNOWN>
+            { extraPartitionOption = SqlPartitionByNoneUnknown.UNKNOWN; }
+        )
+    ]
+    <RPAREN>
+    {
+        return new SqlRangeN(getPos(), testExpression, rangeList,
+            extraPartitionOption);
+    }
+}
+
+SqlRangeNStartEnd RangeNStartEnd() :
+{
+    SqlNode startLiteral = null;
```

Comment:
nit: Could this be made final since it is always parsed

Reply:
Fixed in https://github.com/googleinterns/calcite/pull/212/commits/596f9761b3fc725a5599e09f123cae7aafa6d4a4